Out-of-Core Hash Aggregate

lnkuiper opened the pull request:
This PR implements out-of-core functionality for our aggregate hash table.

This is done by abandoning hash tables when they reach a specific memory limit in `PartitionableHashTable::ListAddChunk`, as well as repartitioning hash tables in `RadixPartitionedHashTable::ScheduleTasks` partitions would otherwise not fit within the memory limit.

This is a basic implementation, and I have many ideas to improve our hash aggregate, which I will implement in the coming weeks. For now, though, I wanted to send this PR with the out-of-core functionality so this PR is manageable.

This implementation also forces all aggregates that use an `Allocator` to use `ArenaAllocator` instead. This means that aggregate states are never explicitly destroyed but are destroyed when the `ArenaAllocator` goes out of scope. We already used the `Allocator` wrapper for the `ArenaAllocator`, so the aggregate states were usually not destroyed explicitly. Of course, we want to destroy/serialize the aggregate states for the out-of-core hash aggregate. I will address this at some point in the future, but I will focus on improving the hash aggregate (both in-memory and out-of-core) first.

As always, I have a little benchmark! This is the query:
```sql
SELECT count(*) FROM (SELECT DISTINCT * FROM 'lineitem30.parquet');
```
Where `'lineitem30.parquet'` is the first 30M rows from TPC-H `lineitem` at SF10. These are the results:

|  memory limit | runtime   |
|:-------------:|:---------:|
| 10.0GB        | 8.52s     |
| 9.0GB         | 9.35s     |
| 8.0GB         | 7.89s     |
| 7.0GB         | 7.81s     |
| 6.0GB         | 6.84s     |
| 5.0GB         | 7.39s     |
| 4.0GB         | 9.04s     |
| 3.0GB         | 8.76s     |
| 2.0GB         | 8.08s     |
| 1.0GB         | 10.27s    |

Interestingly, the execution time decreases as the memory limit decreases. This is due to the out-of-core strategy creating more fine-grained partitions. This improves cache-locality, which explains the improved runtime.

Commit:
use BufferAllocator in a bunch of places, and reduce memory pressure in Agg HT partitioning

Commit:
initial version of external hash aggregate

Commit:
add out-of-core verification to aggregate tests

Commit:
out-of-core hash aggregate working mostly

Commit:
Merge branch 'feature' into oocha

Commit:
merge with feature

Commit:
move to ArenaAllocator for all aggregates that allocate

Commit:
refactor to always use ArenaAllocator in aggregates

Commit:
Merge branch 'feature' into oocha

Commit:
merge with feature branch

Commit:
abandon HT at memory threshold

Commit:
Merge branch 'feature' into oocha

Commit:
Merge branch 'feature' into oocha

Commit:
Merge branch 'feature' into oocha

Commit:
Merge branch 'feature' into oocha

Commit:
check whether we've partitioned before checking memory limit

Commit:
Merge branch 'feature' into oocha

Commit:
properly arena-allocate aggregates

Commit:
merge with feature

Commit:
fix radix bit count logic

Commit:
rework so out-of-core aggregate does not spinlock, and add another test

Commit:
Merge branch 'feature' into oocha

Commit:
fix leak

hawkfish commented on test/api/test_reset.cpp:
```diff
@@ -90,8 +90,7 @@ OptionValueSet &GetValueForOption(const string &name) {
 	    {"arrow_large_buffer_size", {true}},
-	    {"allocator_flush_threshold", {"4.2GB"}}
-	};
+	    {"allocator_flush_threshold", {"4.2GB"}}};
```
I think the tidy tasks are complaining about this?

hawkfish commented on src/execution/radix_partitioned_hashtable.cpp:
```diff
@@ -302,22 +323,244 @@ class RadixAggregateFinalizeTask : public ExecutorTask {
+		} else {
+			// Repartition to keep all threads busy
+			repartition_radix_bits = RadixPartitioning::RadixBits(NextPowerOfTwo(n_threads));
+			tasks_per_partition = n_threads / num_partitions;
```
If n_threads == num_partitions + 1 then won't a thread be idle?

lnkuiper replied:
No, then it will start repartitioning the next partition

Mytherin commented on test/sql/aggregate/simple_external_aggregate.test_slow:
```diff
@@ -0,0 +1,34 @@
+----
+1000000
+
+mode skip
```
Is this left in on purpose?

lnkuiper replied:
Nope, oops!

Mytherin commented on src/execution/radix_partitioned_hashtable.cpp:
```diff
@@ -163,12 +174,15 @@ void RadixPartitionedHashTable::Sink(ExecutionContext &context, DataChunk &chunk
-		                                      group_types, op.payload_types, op.bindings);
+		    make_uniq<PartitionableHashTable>(context.client, BufferAllocator::Get(context.client),
+		                                      *gstate.partition_info, group_types, op.payload_types, op.bindings);
+		if (context.client.config.force_external) {
```
Not directly related but can we perhaps rename this to `verify_external` in the codebase as well to prevent confusion?

Mytherin commented on src/common/types/list_segment.cpp:
```diff
@@ -668,7 +619,6 @@ void GetSegmentDataFunctions(ListSegmentFunctions &functions, const LogicalType
 		functions.write_data = WriteDataToStructSegment;
 		functions.read_data = ReadDataFromStructSegment;
 		functions.copy_data = CopyDataFromStructSegment;
-		functions.destroy = DestroyStructSegment;
```
Is this required for the repartitioning to work? If so - should all aggregates with destructors have their destructors removed in order for this to work? Can we perhaps add large tests with `verify_external` for all aggregates with destructors (including e.g. `min/max/first` on non-inlined string values, `string_agg`, `histogram`, etc).

Mytherin reviewed (commented):
Thanks for the PR! Looks great - and great performance results! Some comments:

Commit:
Merge branch 'feature' into oocha

Commit:
add verify_external to a bunch of tests

Commit:
add large external aggregate test

Commit:
Merge branch 'feature' into oocha

Commit:
fix wrong arithmetic

Commit:
remove verification from test (breaks)

lnkuiper commented on test/sql/aggregate/group/test_group_by_parallel.test_slow:
```diff
@@ -7,6 +7,9 @@ require skip_reload
 PRAGMA threads=4
 
+statement ok
+PRAGMA verify_external
```
Good idea, I've added a big test case

lnkuiper commented:
Thanks for the PR feedback, both of you! I've added a bunch of `verify_external`, and I've added a large aggregate test with nested types. Due to adding this test, I've had to merge with master because I needed the fix of #7947 to get the test to work.

Commit:
disable_verify_external for reproducibility with setseet

Commit:
Merge branch 'feature' into oocha

Commit:
merge with feature

hawkfish reviewed (commented):
Most of this is in my full vectorisation PR that I just submitted. It was a moderate sized change, so there would be conflicts.

Commit:
Merge branch 'feature' into oocha

Commit:
trying to please the CI

Commit:
merge with feature and fix test

Commit:
short-circuit if forcing external

Commit:
Merge branch 'feature' into oocha

Commit:
improve coverage

Commit:
Merge branch 'feature' into oocha

lnkuiper commented:
I don't think the failing test has anything to do with my changes. This fails on vector size 2, somehow strings become invalid:
```sql
statement ok
CREATE VIEW strlists AS SELECT * FROM (VALUES
	(1, ['a']),
	(2, [NULL]),
	(3, []),
	(4, ['Branta Canadensis', 'c']),
	(5, ['i','j','k']),
	(NULL::INTEGER, ['Somateria mollissima'])
	) lv(pk, p);

query II
SELECT * FROM strlists
----
NULL	[Somateria mollissima]
1	[a]
2	[NULL]
3	[]
4	[Branta Canadensis, c]
5	[i, j, k]
```

Actual result:
```sql
1	[a]
2	[NULL]
3	[]
4	[Branta Canadensis, c]
5	[i, j, Z]
NULL	[ZZZZZZZZZZZZZZZZZZZZ]
```

I will investigate, but maybe I can add a `require vector_size`?

Mytherin commented:
I have not seen this issue before in other PRs - perhaps you could investigate?

Commit:
fix list(varchar) columndatacollection bug

Commit:
fix bug with multiple slice of same selection vector

Commit:
increment uncovered lines to fix spurious failure

Commit:
merge with feature

Commit:
undo vector size change

Commit:
Merge branch 'feature' into oocha

Commit:
properly compute offset when unswizzling list of string

Commit:
Merge branch 'feature' into oocha

Commit:
add db so we can stay in memory limit for sure

lnkuiper commented:
I think this is ready to go!

Mytherin commented:
Thanks! LGTM
